### PR TITLE
Update Developer Guide

### LINKED ISSUE #108: Update Developer's Guide design components
Components to update:

1. Architecture
2. UI component
3. Logic component
4. Model component
5. Storage component
6. Common classes

### PR DESCRIPTION
**Pull request by WadeForward**
Update Developer Guide, mainly with Architecture segment and related diagrams.
Add implementation of `Assign Feature` and related diagrams.

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -64,9 +71,14 @@ The sections below give more details of each component.
 **API** :
 [`Ui.java`](https://github.com/se-edu/addressbook-level3/tree/master/src/main/java/seedu/address/ui/Ui.java)
```
Comment (xz0127) [inline docs/DeveloperGuide.md]: Should the link here be changed?

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -64,9 +71,14 @@ The sections below give more details of each component.
...
+The UI consists of a `MainWindow` that is made up of parts e.g.`CommandBox`, `ResultDisplay`, 
+`PersonListPanel`, `AppointmentListPanel`, `StatusBarFooter` etc. All these, including the `MainWindow`, 
+inherit from the abstract `UiPart` class.
 
-The `UI` component uses JavaFx UI framework. The layout of these UI parts are defined in matching `.fxml` files that are in the `src/main/resources/view` folder. For example, the layout of the [`MainWindow`](https://github.com/se-edu/addressbook-level3/tree/master/src/main/java/seedu/address/ui/MainWindow.java) is specified in [`MainWindow.fxml`](https://github.com/se-edu/addressbook-level3/tree/master/src/main/resources/view/MainWindow.fxml)
+The `UI` component uses JavaFx UI framework. The layout of these UI parts are defined in matching 
+`.fxml` files that are in the `src/main/resources/view` folder. For example, the layout of the 
+[`MainWindow`](https://github.com/se-edu/addressbook-level3/tree/master/src/main/java/seedu/address/ui/MainWindow.java)
```
Comment (xz0127) [inline docs/DeveloperGuide.md]: Should the links on these 2 lines be changed accordingly?

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -97,17 +109,22 @@ Given below is the Sequence Diagram for interactions within the `Logic` componen
...
+* stores the patient book data and appointment book data.
+* exposes an unmodifiable `ObservableList<Patient>` and an unmodifiable `ObservableList<Appointment>` that can be 'observed' 
+e.g. the UI can be bound to these lists so that the UI automatically updates when the data in the list change.
 * does not depend on any of the other three components.
 
 
-<div markdown="span" class="alert alert-info">:information_source: **Note:** An alternative (arguably, a more OOP) model is given below. It has a `Tag` list in the `PatientBook`, which `Person` references. This allows `PatientBook` to only require one `Tag` object per unique `Tag`, instead of each `Person` needing their own `Tag` object.<br>
+<div markdown="span" class="alert alert-info">:information_source: **Note:** 
+An alternative (arguably, a more OOP) Patient model is given below. It has a `Tag` list in the `PatientBook`, 
+which `Person` references. This allows `PatientBook` to only require one `Tag` object per unique `Tag`,
```
Comment (xz0127) [inline docs/DeveloperGuide.md]: Should the `Person` on these two lines be refactored into `Patients`? `Person` is not wrong in this case. What do you think?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -211,11 +228,64 @@ The following activity diagram summarizes what happens when a user executes a ne
...
-### \[Proposed\] Data archiving
+### \[Proposed\] Assign Feature
+
+#### Proposed Implementation
+
+The Assign feature is implemented to allow users to assign a specified patient into a specified appointment date and time. 
+To avoid cyclic dependency, only an `Appointment` has an attribute of `Patient` object instead of `Appointment` object and 
+`Patient` object refer to each other.<br><br>
+
+This feature creates a new Appointment instance, which is stored in an instance of `UniqueAppointmentBook`, which in turn is stored
```
Comment (xz0127) [inline docs/DeveloperGuide.md]: Should it be `UniqueAppointmentList` instead of `UniqueAppointmentBook`?

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -211,11 +228,64 @@ The following activity diagram summarizes what happens when a user executes a ne
...
+#### Proposed Implementation
+
+The Assign feature is implemented to allow users to assign a specified patient into a specified appointment date and time. 
+To avoid cyclic dependency, only an `Appointment` has an attribute of `Patient` object instead of `Appointment` object and 
+`Patient` object refer to each other.<br><br>
+
+This feature creates a new Appointment instance, which is stored in an instance of `UniqueAppointmentBook`, which in turn is stored 
+in the `AppointmentBook`. These classes are part of the `model` component.<br><br>
+
+The feature is supported by the AssignCommand class which extends the abstract class `Command`, and `AssignmentCommandParser`
```
Comment (xz0127) [inline docs/DeveloperGuide.md]: Would it be better if you add `` to AssignCommand?

**Review by xz0127 — COMMENTED** [review]
Other than the comments mentioned, the rest LGTM! 👍

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (WadeForward) [reply docs/DeveloperGuide.md]: Oh, this UI part has been done by Yang Yue, so I guess this one can just ignore.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (WadeForward) [reply docs/DeveloperGuide.md]: Good spot!

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (WadeForward) [reply docs/DeveloperGuide.md]: Oh this part is supposed to be Ming Lim's, I accidentally touched it.

**Review by JinHao-L — APPROVED** [review]
LGTM! 👍

**Review by JinHao-L — APPROVED** [review]
Other than the merge conflicts, LGTM! 👍

**Pull request merged**